// Test Data Service - gRPC Service Definition
syntax = "proto3";

package testdata.v1;

// Test Data Service
service TestDataService {
  // Synchronous generation for small requests (<1000 records)
  rpc GenerateData(GenerateRequest) returns (GenerateResponse);

  // Streaming for large requests
  rpc GenerateDataStream(GenerateRequest) returns (stream DataChunk);

  // List available schemas
  rpc GetSchemas(GetSchemasRequest) returns (GetSchemasResponse);

  // Health check
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// Request to generate test data
message GenerateRequest {
  string request_id = 1;                    // Unique ID for tracing
  string domain = 2;                        // ecommerce, supply_chain, loyalty, etc.
  string entity = 3;                        // cart, order, payment, user, review
  Schema schema = 4;                        // Field definitions (or use pre-defined)
  Constraints constraints = 5;              // Validation rules
  repeated Scenario scenarios = 6;          // Generation scenarios
  string context = 7;                       // Natural language context for LLM
  repeated string hints = 8;                // Routing hints: realistic, edge_case, defect_triggering
  OutputFormat output_format = 9;           // JSON, CSV, SQL
  int32 count = 10;                         // Total records to generate
  bool use_cache = 11;                      // Use cached data pools
  bool learn_from_history = 12;             // Use RAG patterns
  bool defect_triggering = 13;              // Generate bug-finding data
  bool production_like = 14;                // Mimic production distributions
  string inline_schema = 15;                // JSON string of complete schema definition
}

// Schema definition
message Schema {
  repeated Field fields = 1;
  string predefined_schema = 2;             // Use pre-defined: "cart", "order", etc.
}

// Field definition
message Field {
  string name = 1;
  FieldType type = 2;
  bool required = 3;
  string description = 4;                   // For LLM context
  repeated Field nested_fields = 5;         // For object/array types
}

// Field types
enum FieldType {
  STRING = 0;
  INTEGER = 1;
  FLOAT = 2;
  BOOLEAN = 3;
  DATE = 4;
  DATETIME = 5;
  EMAIL = 6;
  PHONE = 7;
  ADDRESS = 8;
  UUID = 9;
  ENUM = 10;
  OBJECT = 11;
  ARRAY = 12;
}

// Constraints for validation
message Constraints {
  map<string, FieldConstraint> field_constraints = 1;
}

// Constraint for a specific field
message FieldConstraint {
  optional double min = 1;
  optional double max = 2;
  repeated string enum_values = 3;
  optional string regex = 4;
  optional int32 min_length = 5;
  optional int32 max_length = 6;
  optional string format = 7;               // date format, etc.
}

// Test scenario definition
message Scenario {
  string name = 1;                          // happy_path, expired_token, etc.
  int32 count = 2;                          // Records for this scenario
  map<string, string> overrides = 3;        // Field value overrides
  string description = 4;                   // For LLM context
}

// Output formats
enum OutputFormat {
  JSON = 0;
  CSV = 1;
  SQL = 2;
}

// Response with generated data
message GenerateResponse {
  string request_id = 1;
  bool success = 2;
  string data = 3;                          // JSON string of generated data
  int32 record_count = 4;
  GenerationMetadata metadata = 5;
  string error = 6;
}

// Metadata about the generation
message GenerationMetadata {
  string generation_path = 1;               // traditional, llm, rag, hybrid
  int32 llm_tokens_used = 2;
  float generation_time_ms = 3;
  float coherence_score = 4;
  map<string, int32> scenario_counts = 5;
}

// Data chunk for streaming
message DataChunk {
  string request_id = 1;
  string data = 2;                          // Partial JSON array
  int32 chunk_index = 3;
  bool is_final = 4;
}

// Request to list schemas
message GetSchemasRequest {
  string domain = 1;                        // Filter by domain (optional)
}

// Response with available schemas
message GetSchemasResponse {
  repeated SchemaInfo schemas = 1;
}

// Information about a schema
message SchemaInfo {
  string name = 1;
  string domain = 2;
  string description = 3;
  repeated string fields = 4;
}

// Health check request
message HealthCheckRequest {}

// Health check response
message HealthCheckResponse {
  string status = 1;                        // healthy, degraded, unhealthy
  map<string, string> components = 2;       // Component health status
}
